src/uct/ib/base/ib_device.c

```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
     *numa_node = (status == UCS_OK) ? n : -1;
 }
 
+static void
+uct_ib_device_async_event_dispatch(uct_ib_device_t *dev,
+                                   uct_ib_async_event_t *event)
+{
+    uct_ib_async_event_val_t *entry;
+    khiter_t iter;
+    int ret;
+
+    ucs_spin_lock(&dev->async_event_lock);
+    iter = kh_put(uct_ib_async_event, &dev->async_events_hash, *event, &ret);
+    entry = &kh_value(&dev->async_events_hash, iter);
+    if (ret == 0) {
+        if (entry->cb != NULL) {
```

Comment:
in which case we can have entry with NULL callback?

Reply:
event before subscription (i.e. before tx flushed)

Reply:
but then i think we would not have any entry at all?

Reply:
i mean event arrived before. flag=1, no subscription yet.